### ISSUE TITLE
How to create single request object from multiple queries input arguments?

### ISSUE THREAD

**Issue by TechGeeky** [graphql-dotnet/graphql-dotnet#2496]
I am working on migrating legacy use cases to our new `GraphQL` service. 

- Internally all our requests gets mapped to single `ProcessRequest` object which holds all input arguments being passed from Schema. 
- Then we pass this `ProcessRequest` object to underlying methods which in turn returns back `ProcessResponse` object which has response for that request.

Below is `ProcessRequest` class:

    public interface IProcess
    {
        int ClientId { get; set; }
        int LanguageId { get; set; }
        int CustomerId { get; set; }
    }

    public interface IData
    {
        IList<Filter> Filters { get; set; }
        KeywordsTemplate KeywordsTemplate { get; set; }
    }

    public class Filter
    {
        public string Key { get; set; }
        public int Value { get; set; }
    }

    public class KeywordsTemplate
    {
        public string Query { get; set; }
        public string UserQuery { get; set; }
    }

    public class ClientType
    {
        public string ClientKey { get; set; }
    }

    public class Shop
    {
        public double CartAmount { get; set; }
        public double PromoAmount { get; set; }
        public IList<int> ClientIds { get; set; }
    }

    public class ListingRequest
    {
        public IList<Filter> Filters { get; set; }
        public int ClientNumber { get; set; }
        public string PotType { get; set; }
        public IList<int> ClientIds { get; set; }
    }

    public class ListingsRequest
    {
        public IList<ListingRequest> Listings { get; set; }
    }

    public class ProcessRequest : IData, IProcess
    {
        public DetailsRequest DetailsRequest { get; set; }
        public ClientType ClientType { get; set; }
        public ListingsRequest ListingsRequest { get; set; }
        public ProtoRequest ProtoRequest { get; set; }

        public IList<Filter> Filters { get; set; }
        public KeywordsTemplate KeywordsTemplate { get; set; }
        public int ClientId { get; set; }
    }

For my below `GraphQL` query - I map input arguments to my `ProcessRequest` class and then call corresponding methods to get `ProcessResponse` object. Then I created a corresponding model `LandingGroup` from `ProcessResponse` class which exactly matches with what I need in `GraphQL` response for my below `contentRequest` query and then I use that in my `ContentResponseType` class to get what I need.

    {
      contentRequest(clientId: 123, filters: [{key: "abc", value: "98"}], clientKey: "process") {
        title
        id
        name
      }
    }

Below is how I am using my Query -

    public class TitanQuery : ObjectGraphType<object>
    {
        public TitanQuery(....)
        {
            Name = "Query";

            FieldAsync<ListGraphType<ContentResponseType>>(
                "contentRequest",
                arguments: new QueryArguments(
                        new QueryArgument<NonNullGraphType<IntGraphType>>
                        {
                            Name = "clientId",
                            Description = "Client Id of the request"

                        },
                        new QueryArgument<ListGraphType<FilterType>>
                        {
                            Name = "filters",
                            Description = "Filters"
                        },
                        new QueryArgument<StringGraphType>
                        {
                            Name = "clientKey",
                            DefaultValue = "abc",
                            Description = "some stuff"
                        }
                    ),

                resolve: async context =>
                {
                    var clientId = context.GetArgument<int>("clientId");
                    var filters = context.GetArgument<List<Filter>>("filters");
                    var clientKey = context.GetArgument<string>("clientKey");

                    // mapping all the arguments to ProcessRequest class
                    ProcessRequest processRequest = new ProcessRequest()
                    {
                        ClientId = clientId,
                        Filters = filters,
                        ClientType = new ClientType()
                        {
                            ClientKey = clientKey
                        },
                        ProtoRequest = new ProtoRequest()
                    };

                    // below line returns 'ProcessResponse' object by using 'processRequest' object
                    var processResponse = await metadataService.ProcessData(processRequest);

                    var dataGroups = processResponse.LandingResponse.DataGroups.Select(dataGroup => new LandingGroup
                    {
                        DataGroup = dataGroup,
                        TileContext = processResponse.TileContext
                    });

                    return dataGroups;
                }
            );
        }
    }

**Problem Statement**

Everything works fine for my above single schema use case. Now I have a very different use case with multiple schemas as shown below. I am not able to figure out on how to do it or whether it is even possible to do it? As you can see below I have three different schemas. 

- All the input arguments from below queries maps to single `ProcessRequest` object. 
- And then I get single `ProcessResponse` object back from `ProcessRequest` object like the same old way as shown above in the `TitanQuery` class.
- All the GraphQL response fields for each schema is already there in `ProcessResponse` object so I need to find a way to return all those fields back to `GraphQL` response for each query.
- `clientId`, `filters` will be there in all the schemas with same values. They will map to corresponding field in `ProcessRequest` class.

Below is my query with multiple schemas:

    {
      contentRequest(clientId: 123, filters: [{key: "hello", value: "98"}], clientKey: "process") {
        title
        id
        name
      }
      find(queryString: "", filters: [{key: "hello", value: "98"}], clientId: 1234) {
        metricx {
          name
          key
        }
        prodCount
      }
      itemListings(pots: [{potType: "KITE_1", filters: [{key: "c", value: "1341"}]}, {potType: "KITE_2", filters: [{key: "c", value: "9812"}]}], filters: [{key: "hello", value: "98"}], clientId: 1234, clientKey: "process") {
        potType
        productAlgo
        count
        listingEndpoint
      }
    }

Now my question is how can I make single `ProcessRequest` object by extracting input arguments from all my above queries? And then from single `ProcessResponse` object I need to extract all those fields which each of my query needs in `GraphQL` response. Also I am not sure if we can have different models from `ProcessResponse` object for different combinations of schemas and use them?

**Note:**

- I have simplified my response fields for each query to make it easier for understanding but in general they are complicated with lot of fields but all of those fields for each query are part of `ProcessResponse` object only.
- Also this is our legacy use case which we are migrating to our new `GraphQL` service and that legacy service passes queries in that format as of now so we are trying to follow the same format if we can.

**Comment by TechGeeky** [comment]
@Shane32 Interested in hearing your thoughts to see if this is possible to achieve it?

**Comment by Shane32** [comment]
Sure it is possible.  I have something similar in my own implementation.

For the sake of discussion, and since it will be easier to implement, let's add a layer to the query:

```gql
{
  multipleQueryRequest {
    contentRequest(clientId: 123, filters: [{key: "hello", value: "98"}], clientKey: "process") {
      title
      id
      name
    }
    find(queryString: "", filters: [{key: "hello", value: "98"}], clientId: 1234) {
      metricx {
        name
        key
      }
      prodCount
    }
    itemListings(pots: [{potType: "KITE_1", filters: [{key: "c", value: "1341"}]}, {potType: "KITE_2", filters: [{key: "c", value: "9812"}]}], filters: [{key: "hello", value: "98"}], clientId: 1234, clientKey: "process") {
      potType
      productAlgo
      count
      listingEndpoint
    }
  }
}
```

Then you need to write code in the resolver method of `multipleQueryRequest` to examine all of the requested child fields and formulate a response.  The child graph objects then just pull from their source objects, ignoring the arguments passed to them. 
 It ain't pretty, but it's possible.  I've written a sample to demonstrate:

```cs
    public class Test123Child
    {
        public int Child1a { get; set; }
    }

    public class Test123ChildGraphType : ObjectGraphType<Test123Child>
    {
        public Test123ChildGraphType()
        {
            Field(x => x.Child1a);
        }
    }
    public class Test123
    {
        public Test123Child Child1 { get; set; }
        public Test123Child Child2 { get; set; }
    }
    public class Test123GraphType : ObjectGraphType<Test123>
    {
        public Test123GraphType()
        {
            Field<Test123ChildGraphType, Test123Child>("Child1")
                .Argument<IntGraphType>("arg");
            Field<Test123ChildGraphType, Test123Child>("Child2")
                .Argument<IntGraphType>("arg");
        }
    }
    public class MyQuery : ObjectGraphType
    {
        public MyQuery()
        {
            Field<Test123GraphType>(
                "test123",
                resolve: ctx =>
                {
                    var thisGraphType = (Test123GraphType)ctx.FieldDefinition.ResolvedType.GetNamedType();

                    var test123 = new Test123();
                    var children = ctx.SubFields;
                    if (children.ContainsKey("child1"))
                    {
                        var child = children["child1"];
                        var args = thisGraphType.GetField("child1").Arguments;
                        var arguments = GraphQL.Execution.ExecutionHelper.GetArgumentValues(args, child.Arguments, ctx.Variables);
                        var val = arguments.TryGetValue("arg", out var argumentValue) ? (int)(argumentValue.Value ?? 0) : 0;
                        test123.Child1 = new Test123Child { Child1a = val };
                    }
                    if (children.ContainsKey("child2"))
                    {
                        var child = children["child2"];
                        var args = thisGraphType.GetField("child2").Arguments;
                        var arguments = GraphQL.Execution.ExecutionHelper.GetArgumentValues(args, child.Arguments, ctx.Variables);
                        var val = arguments.TryGetValue("arg", out var argumentValue) ? (int)(argumentValue.Value ?? 0) : 0;
                        test123.Child2 = new Test123Child { Child1a = val };
                    }
                    return test123;
                });
        }
    }

        [Fact]
        public void RunTest()
        {
            var queryString = "query($arg: Int! = 2, $arg2: Int! = 3) { test123 { child1(arg: $arg) { child1a } child2(arg: $arg2) { child1a } } }";
            var expectedAnswer = @"{ ""test123"": { ""child1"": { ""child1a"": 2 }, ""child2"": { ""child1a"": 3 } } }";
            AssertQuerySuccess(queryString, expectedAnswer, null);
        }
```

As you can see, most of the helper methods are unavailable to you because there is no `IResolveFieldContext` instance for the child fields.  Moreover, trying to examine the children of those children is still harder, because `SubFields` is not available.  And this implementation does not handle requests with aliases properly.  For instance, this request is entirely valid:

```gql
{
  multipleQueryRequest {
    hello: contentRequest(clientId: 123, filters: [{key: "hello", value: "98"}], clientKey: "process") {
      title
      id
      name
    }
    widget: contentRequest(clientId: 123, filters: [{key: "widget", value: "98"}], clientKey: "process") {
      title
      id
      name
    }
  }
}
```

But my sample code does not properly handle this request.  To properly handle this request also, you'd need to change the `Test123` class into a dictionary and store values into it based on the alias.  Then in the resolver of the child, it would need to pull the child object from the source (a dictionary), using the field alias as the dictionary key.  Finally, I didn't check to see if my sample works properly for fragments.

**Comment by Shane32** [comment]
Is this issue related to or a duplicate of #2464 or #2465 @TechGeeky ?

**Comment by TechGeeky** [comment]
I think it will be duplicate of #2464 but we can close both of those for now. This is the main thread where I am working as of now.

Reading your answer trying to understand on how this will fit with my example in the question.

**Comment by TechGeeky** [comment]
@Shane32 Do we really need to add layer to the request like `multipleQueryRequest` you added? In my scenario I can have single query or multiple queries in the request as it's not fixed. Our customers can call us by passing single query in the request or multiple queries in the request so is there any way to do this without using any root layer like `multipleQueryRequest`?

**Comment by Shane32** [comment]
Well sure you can write your own document executer implementation and examine the document before it calls the execution strategy. Then set the root object on the execution context before calling the execution strategy, so the source object is set in the context of the root query object. 

Of course if you keep the extra layer that doesn’t mean it wouldn’t work for a single request.

**Comment by TechGeeky** [comment]
Yeah it will work for a single request for sure. The reason I was asking is because as of now our customers call our *old service* like this for single query request - 

    {
      contentRequest(clientId: 123, filters: [{key: "hello", value: "98"}], clientKey: "process") {
        title
        id
        name
      }
    }

or like this for multiple query request -

    {
      contentRequest(clientId: 123, filters: [{key: "hello", value: "98"}], clientKey: "process") {
        title
        id
        name
      }
      find(queryString: "", filters: [{key: "hello", value: "98"}], clientId: 1234) {
        metricx {
          name
          key
        }
        prodCount
      }
      itemListings(pots: [{potType: "KITE_1", filters: [{key: "c", value: "1341"}]}, {potType: "KITE_2", filters: [{key: "c", value: "9812"}]}], filters: [{key: "hello", value: "98"}], clientId: 1234, clientKey: "process") {
        potType
        productAlgo
        count
        listingEndpoint
      }
    }

Since we migrating them from old service to new service so was trying to make sure they don't have to the change their request at all. This way they can just call new GraphQL endpoint with same requests and everything will work as it is.

Now if we have to change the query request then I need to ask them to wrap their query with another static layer which I will use to extract child objects I guess. I can do that too but just trying to figure out if there is a way to do that without asking them to change anything?

Also *old service* is written in `Javascript` and uses `GraphQL` as well so does `Javascript` allows that to extract child objects without using any extra layer like we are adding here? This is just for my knowledge.

**Comment by Shane32** [comment]
Certainly if you have an existing graphql schema then you would not want to change it.

There's another way you can implement your schema that would probably work much better: use a batch data loader to group up a number of requests and return them to each of the field resolvers.  Here's how I would set it up:

1. Create a dummy interface and classes containing the request parameters each type of query.

```cs
public interface IRequest
{
}

public class ContentRequest : IRequest
{
    public int ClientId { get; set; }
    public List<Filter> Filters { get; set; }
    public string ClientKey { get; set; }
}

public class FindRequest : IRequest
{
    public string QueryString { get; set; }
    // other query parameters
}

public class ItemListingsRequest : IRequest
{
    // see above
}
```

2. Create a custom data loader where the key is the dummy interface, and it returns an object.  The "data loader" can group all of the requests together server-side and respond with data for each of them.  Be sure caching is turned off.

```cs
public class MyDataRequestor : DataLoaderBase<IRequest, object>
{
    private readonly IDataService _dataService; // <-- example service

    public MyDataRequestor(IDataService dataService) : base(false)  // no caching
    {
        // pull DI services as necessary here
        _dataService = dataService;
    }

    protected override async Task FetchAsync(IEnumerable<DataLoaderPair<IRequest, object>> list, CancellationToken cancellationToken)
    {
        // process all of the requests, calling SetResult on each of the pairs
        // below is a simplified example that executes the requests sequentially, but you can change this as desired
        foreach (var pair in list)
        {
            pair.SetResult(pair switch
                {
                    // if the request is a ContentRequest, return a List<ContentResponse>
                    ContentRequest contentRequest => await _dataService.ProcessContentRequestAsync(contentRequest, cancellationToken),

                    FindRequest findRequest => await _dataService.ProcessFindRequestAsync(findRequest, cancellationToken),

                    ItemListingsRequest itemListingsRequest => await _dataService.ProcessItemListingsRequestAsync(itemListingsRequest, cancellationToken),

                    _ => throw new NotSupportedException()
                });
        }
    }
}
```

3. Register the data loader in your DI container as a singleton or scoped, depending on your needs.  If you are going to register it as a singleton, be sure that data loader caching is turned off, or else it will eventually use all of the memory on your server as there is no technique to clear the cache.  Caching won't be useful anyway because without `Equals` being implemented on the class, none of the `IRequest` objects will equal each other.  (It is something that could be implemented, however.)

4. Call the data loader from the `contentRequest`, `find`, and `itemListings` field resolvers.  If the data loader is registered as a singleton, you can include a reference in the graph type constructor.  If not, you will need to obtain a reference from `context.RequestServices`.  I am going to assume it is scoped in the below example.

```cs
public class TitanQuery : ObjectGraphType<object>
{
    public TitanQuery(....)
    {
        Name = "Query";

        FieldAsync<ListGraphType<ContentResponseType>>(
            "contentRequest",
            arguments: new QueryArguments(
                    new QueryArgument<NonNullGraphType<IntGraphType>>
                    {
                        Name = "clientId",
                        Description = "Client Id of the request"

                    },
                    new QueryArgument<ListGraphType<FilterType>>
                    {
                        Name = "filters",
                        Description = "Filters"
                    },
                    new QueryArgument<StringGraphType>
                    {
                        Name = "clientKey",
                        DefaultValue = "abc",
                        Description = "some stuff"
                    }
                ),

            resolve: async context =>
            {
                var dataLoader = context.RequestServices.GetRequiredService<MyDataRequestor>();
                var request = new ContentRequest {
                    ClientId = context.GetArgument<int>("clientId"),
                    Filters = context.GetArgument<List<Filter>>("filters"),
                    ClientKey = context.GetArgument<string>("clientKey"),
                };

                // after being processed by the dataloader, will return a List<ContentResponse>
                return dataLoader.LoadAsync(request);
            }
        );
    }
}
```

FYI, how data loaders work is that they will execute all nodes possible until none remain that can be executed.  Then it will execute/trigger all pending data loader nodes at once.  See `SerialExecutionStrategy` and `ParallelExecutionStrategy` source code for specifics.

Also, I'd be real careful about registering the data loader as a singleton.  Specifically, let's say two request come in at the same time like this:

```gql
# request A

query {
  contentRequest { ... }
}

# request B

query {
  contentRequest { ... }
  find { ... }
}
```

Then what can happen is that the two `contentRequest` queries can get grouped into the same dataloader call, while the second half of request B (the `find` query) gets put into a separate call.  Worse yet, the first group of requests gets called from the context of request A, so if there is any code that relies on `IHttpContextAccessor`, it will return the wrong http context for the second content request. Similar issues can occur with any async-local data.  I had these problems in one of my live environments; it was very hard to track down since the probability of it happening is rare.  Now if that is perfectly fine (no issues with authentication or anything else), then sure, I guess...  But also if the `FetchAsync` code in the data loader depends on any scoped services, you'll need the data loader to be a scoped service.

I think this solution should work well for you.  It also can fully handle any valid graphql query, whether it uses aliases, fragments, etc.

**Comment by TechGeeky** [comment]
Interesting solution. Let me go through this and try to understand. I will get back to you in sometime.

**Comment by TechGeeky** [comment]
You meant `FetchAsync` right in your return line in `TitanQuery` class?

> return dataLoader.LoadAsync(request);

Also as of now it gives me an error on that line - 

`'MyDataRequestor.FetchAsync(IEnumerable<DataLoaderPair<IRequest, object>>, CancellationToken)' is inaccessible due to its protection level`

**Comment by Shane32** [comment]
No, `LoadAsync` is correct.

**Comment by Shane32** [comment]
Here's the gory details:

`LoadAsync` queues the key (the `IRequest` in your case) and returns an `IDataLoaderResult<object>` interface (which is an instance of `DataLoaderPair<IRequest, object>`).  Once the execution strategy determines that only data loader nodes are remaining to be processed, it will call `await IDataLoaderResult.GetResultAsync(cancellationToken)`.  The `DataLoaderPair` in turn holds a reference of the queued requests in the form of a `DataLoaderList` instance and calls the `DispatchAsync` method.  The `DataLoaderList.DispatchAsync` method holds a reference to the `DataLoaderBase` and calls the abstract `FetchAsync` method to load the results into the list for all of the pending nodes.  The `FetchAsync` method contains only your own code and calls the `DataLoaderPair.SetResult` methods once the information requested has been retrieved.  Once the `DispatchAsync` method finishes, the `DataLoaderPair.GetResultAsync` method returns the data back to the execution strategy.

For the other pending data loader nodes, one of two things happen:
1.  If the `DispatchAsync` method has already completed, then the `IsResultSet` property of the pair is already `true`, and so the `GetResultAsync` method can immediately return the correct data.
2. If not, it calls `DispatchAsync`, which returns a reference to the same `Task` that represents the completion of the `FetchAsync` method on the other thread.  Once that task is complete, the `IsResultSet` property will be `true` and the data can be returned from the pair.

**Comment by TechGeeky** [comment]
Yeah my bad. I missed `LoadAsync` in `DataLoaderBase` class. Went through the doc and got some clarity. 

I am kinda confuse on your second solution and not sure how it will fit to what I need unless I am understanding it wrong. This is what I am trying to do and let me know if your second solution helps me with that -

When customer passes query like this - 

    {
      contentRequest(clientId: 123, filters: [{key: "hello", value: "98"}], clientKey: "process") {
        title
        id
        name
      }
    }

then I want to create `ProcessRequest` object like below.

    var processRequest = new ProcessRequest {
        ClientId = context.GetArgument<int>("clientId"),
        Filters = context.GetArgument<List<Filter>>("filters"),
        ClientKey = context.GetArgument<string>("clientKey"),
    };

But if customer passes multiple queries in single request like this - 

    {
      contentRequest(clientId: 123, filters: [{key: "hello", value: "98"}], clientKey: "process") {
        title
        id
        name
      }
      find(queryString: "", filters: [{key: "hello", value: "98"}], clientId: 1234) {
        metricx {
          name
          key
        }
        prodCount
      }
      itemListings(pots: [{potType: "KITE_1", filters: [{key: "c", value: "1341"}]}, {potType: "KITE_2", filters: [{key: "c", value: "9812"}]}], filters: [{key: "hello", value: "98"}], clientId: 1234, clientKey: "process") {
        potType
        productAlgo
        count
        listingEndpoint
      }
    }

then I want to create request object like this by extracting arguments from all those queries and then pass this `ProcessRequest` object to underyling methods to get `ProcessResponse` object. `clientId`, `filters`, `clientKey` will be there in all the schemas with same values mostly.

    var processRequest = new ProcessRequest {
        ClientId =    'clientId from any query arguments',
        Filters =     'filters from any query arguments',
        ClientKey =   'clientKey from any query arguments',
        QueryString = 'queryString from "find" query arguments',
        Pots =        'pots from "itemListings" query arguments'
    };

Basically arguments from each query is mapped to some object or variables in `ProcessRequest` class. Single query or multiple query makes one `ProcessRequest` object. Once I make `processRequest` object then I pass this object to underyling methods of services to get `ProcessResponse` object just like I have in my original question - 

> var processResponse = await metadataService.ProcessData(processRequest);

Once I have `ProcessResponse` object which has all the data either from single query or multiple queries then I need to get corresponding fields for each query type from it.

Now in our code base, we already have `ProcessRequest` and `ProcessResponse` defined. We just need to create `ProcessRequest` from either single query or multiple query arguments and then from that get `ProcessResponse` which will be use to return data for query fields.

Now if I see your example of `FetchAsync` method then it looks like we are creating different input request object for each query arguments but in my case I need single `ProcessRequest` object from all those query arguments and use that to get `ProcessResponse` object.

**Comment by Shane32** [comment]
The NPE is because `ExecutionOptions.RequestServices` isn't set when calling the `DocumentExecutor`.

**Comment by Shane32** [comment]
Yes, but you can change the code in FetchAsync however you want. For example:

```cs
    protected override async Task FetchAsync(IEnumerable<DataLoaderPair<IRequest, object>> list, CancellationToken cancellationToken)
    {
        //1. find the different types of requests, and ensure that only max of one is requested for each type
        //  (because according to graphql spec there could be multiple of a single one)
        var contentRequestPair = list.SingleOrDefault(x => x.Key is ContentRequest);
        var contentRequest = (ContentRequest)contentRequestPair?.Key;
        //ditto for findRequest and itemListingsRequest

        //2. pull the common variables
        var clientId = contentRequest?.ClientId ?? findRequest?.ClientId ?? itemListingsRequest?.ClientId;
        //ditto for filters and clientkey

        //optional: throw an exception if clientId is different for contentRequest vs findRequest vs itemListingsRequest, or similar logic

        //3. initialize the process request
        var request = new ProcessRequest {
            //common properties
            ClientId = clientId,
            ClientKey = clientKey,
            Filters = filters,
            //properties specific to the type of request
            QueryString = findRequest?.Find,
            ...
        };

        // 4. todo: execute your request (async or sync)
        var processResponse = await ExecuteProcessRequestAsync(processRequest, cancellationToken);

        // 5. set the results
        if (contentRequestPair != null)
        {
            // create the data model for the graph
            var contentResponse = processResponse.whatever;
            // set the answer
            contentRequestPair.SetResult(contentResponse);
        }
        listRequestPair?.SetResult(processResponse.answer2);
        // ditto for the other objects
    }
```

Some more idea: group the `list` variable by the common properties, so e.g. if there are different clientIds specified it will execute separate requests rather than execute one of the two with the wrong clientid.  Apply similar logic if there are multiple `contentRequest`s queried, so it will execute separate requests rather than throwing an error during `SingleOrDefault`.

**Comment by Shane32** [comment]
> Do you think DataLoader can still be used here? Maybe FetchAsync method can be modified to make one ProcessRequest object from all input arguments in multiple queries and then call await metadataService.ProcessData(processRequest) to get data and then return multiple response type for each resolvers? Is that possible to do by any chance?

Yes exactly.  This is your responsibility now to implement `FetchAsync` any way you want.  Group the queries together into one call, or issue separate calls, whatever.

**Comment by TechGeeky** [comment]
@Shane32 I think it makes sense now. I got more clarity after your last few comments. Working on making those changes. Also trying to fix the error I am getting by adding this line - `options.RequestServices = context.RequestServices;`

```
        private async Task ExecuteAsync(HttpContext context, ISchema schema)
        {
            var request = Deserialize<GraphQLRequest>(context.Request.Body);
            var result = await _executer.ExecuteAsync(_ =>
            {
                _.Schema = schema;
                _.Query = request?.Query;
                _.OperationName = request?.OperationName;
                _.Inputs = request?.Variables.ToInputs();
                _.RequestServices = context.RequestServices;
                _.UserContext = _settings.BuildUserContext?.Invoke(context);
                _.ValidationRules = DocumentValidator.CoreRules.Concat(new[] { new InputValidationRule() });
                _.EnableMetrics = _settings.EnableMetrics;
                if (_settings.EnableMetrics)
                {
                    _.FieldMiddleware.Use<InstrumentFieldsMiddleware>();
                }
                _.UnhandledExceptionDelegate = ctx => logger.Error("GraphQL", "{Error} occurred", ctx.OriginalException.Message, "GraphQL error", ctx.OriginalException.Message);
            });
            await WriteResponseAsync(context, result);
        }
```

**Comment by sungam3r** [comment]
Since it seems issue is answered I close it.
